s3/pkg/service/object.go

```diff
@@ -116,17 +116,159 @@ func (s *s3Service) PutObject(ctx context.Context, in pb.S3_PutObjectStream) err
+		return ErrInternalError
+	}
+
+	log.Infof("*********bucket:%s,key:%s,size:%d\n", req.BucketName, req.ObjectKey, req.Size)
+	bucket, err := s.MetaStorage.GetBucket(ctx, req.BucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+	if bucket.Versioning.Status == utils.VersioningEnabled {
+		log.Info("Inside PUT Versioned Objects")
+		return PutVersionedObject(ctx, req, in, s)
+	} else {
+		log.Info("Inside PUT  Objects")
+		return PutObject(ctx, req, in, s)
+	}
+}
+
+func PutVersionedObject(ctx context.Context, req *pb.PutObjectRequest, in pb.S3_PutObjectStream, s *s3Service) error {
+	log.Infoln("PutVersionedObject is called in s3 service.")
+
+	var err error
+	result := &pb.PutObjectResponse{}
+	defer func() {
+		result.ErrorCode = GetErrCode(err)
+		in.SendMsg(result)
+	}()
+
 	isAdmin, tenantId, userId, err := util.GetCredentialFromCtx(ctx)
 	if err != nil {
 		log.Errorln("failed to get credential info. err:", err)
 		return nil
 	}
 
-	req := &pb.PutObjectRequest{}
-	err = in.RecvMsg(req)
+	log.Infof("*********bucket:%s,key:%s,size:%d\n", req.BucketName, req.ObjectKey, req.Size)
+	bucket, err := s.MetaStorage.GetBucket(ctx, req.BucketName, true)
 	if err != nil {
-		log.Errorln("failed to get msg with err:", err)
-		return ErrInternalError
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+
+	if !isAdmin {
+		switch bucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if bucket.TenantId != tenantId {
+				err = ErrBucketAccessForbidden
+				return err
+			}
+		}
+	}
+	data := &StreamReader{in: in}
+	var limitedDataReader io.Reader
+	if req.Size > 0 { // request.ContentLength is -1 if length is unknown
+		limitedDataReader = io.LimitReader(data, req.Size)
+	} else {
+		limitedDataReader = data
+	}
+
+	backendName := bucket.DefaultLocation
+	if req.Location != "" {
+		backendName = req.Location
+	}
+	backend, err := utils.GetBackend(ctx, s.backendClient, backendName)
+	if err != nil {
+		log.Errorln("failed to get backend client with err:", err)
+		return err
+	}
+
+	log.Infoln("bucket location:", req.Location, " backendtype:", backend.Type, " endpoint:", backend.Endpoint)
+	bodyMd5 := req.Attrs["md5Sum"]
+	ctx = context.Background()
+	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_SIZE, req.Size)
+	ctx = context.WithValue(ctx, dscommon.CONTEXT_KEY_MD5, bodyMd5)
+	sd, err := driver.CreateStorageDriver(backend.Type, backend)
+	if err != nil {
+		log.Errorln("failed to create storage. err:", err)
+		return err
+	}
+	obj := &pb.Object{BucketName: req.BucketName, ObjectKey: req.ObjectKey}
+	isBucketVersioned := bucket.Versioning.Status == utils.VersioningEnabled
+	if isBucketVersioned {
+		log.Info("Inside service/object.go PUT bucket version enabled")
+		versionId, now := utils.GetVersionId()
+		obj.VersionId = versionId
+		obj.LastModified = now
+		// update the cloud backend object name here
+		obj.ObjectKey = obj.ObjectKey + "_" + obj.VersionId
```

Comment:
VersionId should not be a part of ObjectKey.

Reply:
It's to get the new object key after we enable versioning